### ISSUE TITLE
lint: StatefulWidget could be a StatelessWidget

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/linter#3683]
Occasionally, I see Flutter code where people use StatefulWidgets even though a StatelessWidget would have been sufficient. Example:

```dart
class MyWidget extends StatefulWidget {
  const MyWidget({super.key});

  @override
  State<MyWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

The `State` class doesn't store any state in member variables and it also doesn't use any of the properties provided by `Sate` (e.g. `mounted`). This could easily be refactored to be a more concise `StatelessWidget`:

```dart
class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

I wonder if it would be possible to have a lint for this?

Of course, from a functional perspective using a StatefulWidget here is totally fine, the additional performance overhead of a StatefulWidget over a StatelessWidget is probably neglectable. However, using a StatelessWidget where appropriate leads to more compact, concise, and readable code and is more idiomatic, IMO.

**Comment by bwilkerson** [comment]
> I wonder if it would be possible to have a lint for this?

Yes, the lint appears to be well defined, so it should be possible to implement it. I'm not sure how often code like this is written, so it's hard to gauge how much value such a lint would bring to users, but it seems doable.

**Comment by elbeicktalat** [comment]
> I'm not sure how often code like this is written
@bwilkerson I agree with about the complexity, but if we make the following simple checks than I think we are good to go.

1. Check if widget hasn't a setState
2. Check if the widget hasn't a life cycle methods
3. Check if the widget hasn't any non final field (optional)

**Comment by srawlins** [comment]
@pq as this rule is Flutter-specific, I'm not sure how many "yays" you want in order to declare it "accepted." FWIW there isn't any disagreement.

**Comment by pq** [comment]
Looks good to me. Thanks!

**Comment by Turskyi** [comment]
@pq @srawlins I see this issue has been open for over two years, and the work required to address it was completed, approved, and passing tests over a year ago in [PR #3725 Add unnecessary_stateful_widgets](https://github.com/dart-lang/linter/pull/3725). Unfortunately, due to delays and the repository migration, that contribution was abandoned without resolution.

Could you please provide an update on when we can expect the `unnecessary_stateful_widgets` linter to be available? Given that the delays and lack of communication with the contributor were not their fault, but a process failure, will the team take responsibility for implementing this change in the new repository?

It would be reassuring to see a clear plan to address this issue so that both contributors and users can feel confident that their efforts are valued and respected. Open-source thrives on collaboration, and handling situations like this proactively will help strengthen the community’s trust.

**Comment by srawlins** [comment]
> Could you please provide an update on when we can expect the unnecessary_stateful_widgets linter to be available?

There is no expected availability. It is not actively being worked on.

> Given that the delays and lack of communication with the contributor were not their fault, but a process failure, will the team take responsibility for implementing this change in the new repository?

Sorry for the communication failures. There are no plans to implement this rule by the Dart or Flutter teams.
